// @revideo/2d/lib/components/Media.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { DependencyContext, clamp, isReactive, useLogger, useThread, } from '@revideo/core';
import { computed, initial, nodeName, signal } from '../decorators';
import { Asset } from './Asset';
let Media = class Media extends Asset {
    constructor(props) {
        super(props);
        this.lastTime = -1;
        if (props.play) {
            this.play();
        }
    }
    isPlaying() {
        return this.playing();
    }
    getCurrentTime() {
        return this.clampTime(this.time());
    }
    getDuration() {
        return this.mediaElement().duration;
    }
    getVolume() {
        return this.mediaElement().volume;
    }
    dispose() {
        this.pause();
        this.remove();
        super.dispose();
    }
    completion() {
        return this.clampTime(this.time()) / this.getDuration();
    }
    setCurrentTime(value) {
        const media = this.mediaElement();
        if (media.readyState < 2)
            return;
        media.currentTime = value;
        this.lastTime = value;
        if (media.seeking) {
            DependencyContext.collectPromise(new Promise(resolve => {
                const listener = () => {
                    resolve();
                    media.removeEventListener('seeked', listener);
                };
                media.addEventListener('seeked', listener);
            }));
        }
    }
    setVolume(volume) {
        if (volume < 0 || volume > 1) {
            console.warn(`${volume} is an incorrect value for volume, has to be in range [0,1]. We're clamping to the nearest value`);
        }
        this.mediaElement().volume = Math.min(Math.max(volume, 0), 1);
    }
    setPlaybackRate(playbackRate) {
        let value;
        if (isReactive(playbackRate)) {
            value = playbackRate();
            useLogger().warn({
                message: 'Invalid value set as the playback rate',
                remarks: "<p>The <code>playbackRate</code> of a <code>Video</code> cannot be reactive.</p>\n<p>Make sure to use a concrete value and not a function:</p>\n<pre class=\"wrong\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(<span class=\"hljs-function\">() =&gt;</span> <span class=\"hljs-number\">7</span>);</code></pre><pre class=\"correct\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(<span class=\"hljs-number\">7</span>);</code></pre><p>If you&#39;re using a signal, extract its value before passing it to the property:</p>\n<pre class=\"wrong\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(mySignal);</code></pre><pre class=\"correct\"><code class=\"language-ts\">video.<span class=\"hljs-title function_\">playbackRate</span>(<span class=\"hljs-title function_\">mySignal</span>());</code></pre>",
                inspect: this.key,
                stack: new Error().stack,
            });
        }
        else {
            value = playbackRate;
        }
        this.playbackRate.context.setter(value);
        if (this.playing()) {
            if (value === 0) {
                this.pause();
            }
            else {
                const time = useThread().time;
                const start = time();
                const offset = this.time();
                this.time(() => this.clampTime(offset + (time() - start) * value));
            }
        }
    }
    play() {
        const time = useThread().time;
        const start = time();
        const offset = this.time();
        const playbackRate = this.playbackRate();
        this.playing(true);
        this.time(() => this.clampTime(offset + (time() - start) * playbackRate));
    }
    pause() {
        this.playing(false);
        this.time.save();
        this.mediaElement().pause();
    }
    clampTime(time) {
        const duration = this.getDuration();
        if (this.loop()) {
            time %= duration;
        }
        return clamp(0, duration, time);
    }
    collectAsyncResources() {
        super.collectAsyncResources();
        this.seekedMedia();
    }
    getErrorReason(errCode) {
        let reason;
        switch (errCode) {
            case 1:
                reason = 'MEDIA_ERR_ABORTED';
                break;
            case 2:
                reason = 'MEDIA_ERR_NETWORK. This might be a 404 error.';
                break;
            case 3:
                reason =
                    'MEDIA_ERR_DECODE. This might be an issue with your video file.';
                break;
            case 4:
                reason =
                    'MEDIA_ERR_SRC_NOT_SUPPORTED. If you are sure that the path to the video is correct, this might be a CORS error.';
                break;
            default:
                reason = 'UNKNOWN';
        }
        return reason;
    }
};
__decorate([
    initial(false),
    signal()
], Media.prototype, "loop", void 0);
__decorate([
    initial(1),
    signal()
], Media.prototype, "playbackRate", void 0);
__decorate([
    initial(0),
    signal()
], Media.prototype, "time", void 0);
__decorate([
    initial(false),
    signal()
], Media.prototype, "playing", void 0);
__decorate([
    initial(1),
    signal()
], Media.prototype, "volume", void 0);
__decorate([
    computed()
], Media.prototype, "completion", null);
Media = __decorate([
    nodeName('Media')
], Media);
export { Media };
