// @revideo/2d/lib/components/Rive.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { BBox, useThread, viaProxy, } from '@revideo/core';
import RiveInitializer from '@rive-app/canvas-advanced';
import { computed, initial, nodeName, signal } from '../decorators';
import { Asset } from './Asset';
let Rive = class Rive extends Asset {
    constructor(props) {
        super(props);
        this.currentTime = 0;
        this.lastTime = 0;
        const time = useThread().time;
        const start = time();
        this.time(() => time() - start);
    }
    async rive() {
        const src = viaProxy(this.fullSource());
        const rive = await RiveInitializer({
            locateFile: () => {
                return '/@rive-wasm';
            },
        });
        const canvas = document.createElement('canvas');
        canvas.width = this.width();
        canvas.height = this.height();
        const renderer = rive.makeRenderer(canvas);
        const bytes = await (await fetch(new Request(src))).arrayBuffer();
        const file = (await rive.load(new Uint8Array(bytes)));
        const artboardId = this.artboardId();
        const artboard = this.getArtboard(artboardId, file);
        const animationId = this.animationId();
        const animation = this.getAnimation(animationId, artboard, rive);
        return { rive, renderer, artboard, animation, canvas };
    }
    async draw(context) {
        this.drawShape(context);
        const { rive, renderer, canvas, artboard, animation } = await this.rive();
        const box = BBox.fromSizeCentered(this.computedSize());
        this.currentTime = this.time();
        const timeToAdvance = this.currentTime - this.lastTime;
        this.lastTime = this.currentTime;
        const renderPromise = new Promise(resolve => {
            function renderLoop() {
                renderer.clear();
                animation.advance(timeToAdvance);
                animation.apply(1);
                artboard.advance(timeToAdvance);
                renderer.save();
                renderer.align(rive.Fit.fill, rive.Alignment.center, {
                    minX: 0,
                    minY: 0,
                    maxX: canvas.width,
                    maxY: canvas.height,
                }, artboard.bounds);
                artboard.draw(renderer);
                renderer.restore();
                resolve();
            }
            rive.requestAnimationFrame(renderLoop);
        });
        await renderPromise;
        context.drawImage(canvas, box.x, box.y, box.width, box.height);
        if (this.clip()) {
            context.clip(this.getPath());
        }
        await this.drawChildren(context);
    }
    getArtboard(artboardId, file) {
        if (typeof artboardId === 'string') {
            return file.artboardByName(artboardId);
        }
        if (typeof artboardId === 'number') {
            return file.artboardByIndex(artboardId);
        }
        return file.defaultArtboard();
    }
    getAnimation(animationId, artboard, rive) {
        let animation;
        if (typeof animationId === 'number') {
            animation = artboard.animationByIndex(animationId);
        }
        else if (typeof animationId === 'string') {
            animation = artboard.animationByName(animationId);
        }
        else {
            animation = artboard.animationByIndex(0);
        }
        return new rive.LinearAnimationInstance(animation, artboard);
    }
};
__decorate([
    initial(0),
    signal()
], Rive.prototype, "artboardId", void 0);
__decorate([
    initial(0),
    signal()
], Rive.prototype, "animationId", void 0);
__decorate([
    initial(0),
    signal()
], Rive.prototype, "time", void 0);
__decorate([
    computed()
], Rive.prototype, "rive", null);
Rive = __decorate([
    nodeName('Rive')
], Rive);
export { Rive };
