// @revideo/core/lib/app/bootstrap.js
import DefaultPlugin from '../plugin/DefaultPlugin';
import { setAssetBase } from '../utils';
import { Logger } from './Logger';
import { ProjectMetadata } from './ProjectMetadata';
import { createSettingsMetadata } from './SettingsMetadata';
/**
 * Bootstrap a project.
 *
 * @param name - The name of the project.
 * @param versions - Package versions.
 * @param plugins - Loaded plugins.
 * @param config - Project settings.
 * @param metaFile - The project meta file.
 * @param settingsFile - The settings meta file.
 * @param logger - An optional logger instance.
 *
 * @internal
 */
export function bootstrap(name, versions, plugins, config, metaFile, settingsFile, logger = config.logger ?? new Logger()) {
    const settings = createSettingsMetadata();
    settingsFile.attach(settings);
    const project = {
        name,
        ...config,
        plugins,
        versions,
        settings,
        logger,
        setAssetBase: setAssetBase,
    };
    project.meta = new ProjectMetadata(project);
    project.meta.shared.set(settings.defaults.get());
    project.experimentalFeatures ?? (project.experimentalFeatures = false);
    metaFile.attach(project.meta);
    return project;
}
/**
 * Bootstrap a project together with all editor plugins.
 *
 * @param name - The name of the project.
 * @param versions - Package versions.
 * @param plugins - Loaded plugins.
 * @param config - Project settings.
 * @param metaFile - The project meta file.
 * @param settingsFile - The settings meta file.
 *
 * @internal
 */
export async function editorBootstrap(name, versions, plugins, config, metaFile, settingsFile) {
    const logger = config.logger ?? new Logger();
    const promises = [Promise.resolve(DefaultPlugin())];
    if (config.plugins) {
        for (const plugin of config.plugins) {
            promises.push(parsePlugin(plugin, logger));
        }
    }
    for (const scene of config.scenes) {
        if (scene.plugins) {
            for (const plugin of scene.plugins) {
                promises.push(parsePlugin(plugin, logger));
            }
        }
    }
    for (const plugin of plugins) {
        promises.push(parsePlugin(plugin, logger));
    }
    const pluginSet = new Set();
    const resolvedPlugins = await Promise.all(promises);
    const includedPlugins = [];
    let resolvedConfig = config;
    for (const plugin of resolvedPlugins) {
        if (!plugin || pluginSet.has(plugin.name)) {
            continue;
        }
        pluginSet.add(plugin.name);
        includedPlugins.push(plugin);
        resolvedConfig = {
            ...resolvedConfig,
            ...(plugin.settings?.(resolvedConfig) ?? {}),
        };
    }
    const project = bootstrap(name, versions, includedPlugins, resolvedConfig, metaFile, settingsFile);
    includedPlugins.forEach(plugin => plugin.project?.(project));
    return project;
}
async function parsePlugin(plugin, logger) {
    if (typeof plugin === 'string') {
        try {
            let url = `/@id/${plugin}`;
            const version = new URL(import.meta.url).searchParams.get('v');
            if (version) {
                url += `?v=${version}`;
            }
            plugin = (await import(/* webpackIgnore: true */ /* @vite-ignore */ url)).default();
        }
        catch (e) {
            console.error(e);
            logger.error({
                message: `Failed to load plugin "${plugin}": ${e.message}.`,
                stack: e.stack,
                remarks: e.remarks,
            });
            return null;
        }
    }
    return plugin;
}
