// @revideo/ffmpeg/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVideoCodec = exports.getSampleRate = exports.checkForAudioStream = exports.mergeAudioWithVideo = exports.doesFileExist = exports.getVideoDuration = exports.createSilentAudioFile = exports.concatenateMedia = exports.makeSureFolderExists = exports.resolvePath = void 0;
const ffmpeg = require("fluent-ffmpeg");
const fs = require("fs");
const os = require("os");
const path = require("path");
const uuid_1 = require("uuid");
const settings_1 = require("./settings");
function resolvePath(output, assetPath) {
    let resolvedPath;
    if (assetPath.startsWith('http://') || assetPath.startsWith('https://')) {
        resolvedPath = assetPath;
    }
    else {
        resolvedPath = path.join(output, '../public', assetPath);
    }
    return resolvedPath;
}
exports.resolvePath = resolvePath;
async function makeSureFolderExists(folderPath) {
    if (await fs.promises
        .access(folderPath)
        .then(() => false)
        .catch(() => true)) {
        await fs.promises.mkdir(folderPath, { recursive: true });
    }
}
exports.makeSureFolderExists = makeSureFolderExists;
async function concatenateMedia(files, outputFile) {
    const tempFile = path.join(os.tmpdir(), `${(0, uuid_1.v4)()}.txt`);
    const fileContent = files
        .map(file => `file '${file.replace(/'/g, "\\'")}'`)
        .join('\n');
    await fs.promises.writeFile(tempFile, fileContent);
    return new Promise((resolve, reject) => {
        ffmpeg.setFfmpegPath(settings_1.ffmpegSettings.getFfmpegPath());
        const ffmpegCommand = ffmpeg();
        ffmpegCommand
            .input(tempFile)
            .inputOptions([
            '-f concat',
            '-safe 0',
            '-protocol_whitelist file,http,https,tcp,tls',
        ])
            .outputOptions(['-c copy'])
            .on('error', err => {
            console.error('Error:', err);
            fs.promises.unlink(tempFile).catch(console.error);
            reject(err); // Reject the promise on error
        })
            .on('end', () => {
            fs.promises.unlink(tempFile).catch(console.error);
            resolve(); // Resolve the promise on successful completion
        })
            .save(outputFile);
    });
}
exports.concatenateMedia = concatenateMedia;
async function createSilentAudioFile(filePath, duration) {
    ffmpeg.setFfmpegPath(settings_1.ffmpegSettings.getFfmpegPath());
    return new Promise((resolve, reject) => {
        ffmpeg()
            .addInput(`anullsrc=channel_layout=stereo:sample_rate=${48000}`)
            .inputFormat('lavfi')
            .duration(duration)
            .on('end', () => {
            resolve(filePath);
        })
            .on('error', err => {
            console.error('Error creating silent audio file:', err);
            reject(err);
        })
            .save(filePath);
    });
}
exports.createSilentAudioFile = createSilentAudioFile;
async function getVideoDuration(filePath) {
    ffmpeg.setFfprobePath(settings_1.ffmpegSettings.getFfprobePath());
    return new Promise((resolve, reject) => {
        ffmpeg.ffprobe(filePath, (err, metadata) => {
            if (err) {
                reject(err);
                return;
            }
            const duration = metadata.format.duration;
            if (duration) {
                resolve(duration);
            }
            else {
                reject(new Error('Could not determine video duration.'));
            }
        });
    });
}
exports.getVideoDuration = getVideoDuration;
async function doesFileExist(filePath) {
    try {
        await fs.promises.access(filePath, fs.constants.F_OK);
        return true;
    }
    catch {
        return false;
    }
}
exports.doesFileExist = doesFileExist;
async function mergeAudioWithVideo(audioPath, videoPath, outputPath) {
    ffmpeg.setFfmpegPath(settings_1.ffmpegSettings.getFfmpegPath());
    return new Promise((resolve, reject) => {
        ffmpeg()
            .input(videoPath)
            .input(audioPath)
            .outputOptions(['-c:v', 'copy', '-c:a', 'aac', '-strict', 'experimental'])
            .on('end', () => {
            resolve();
        })
            .on('error', err => {
            console.error(`Error merging video and audio: ${err.message}`);
            reject(err);
        })
            .save(outputPath);
    });
}
exports.mergeAudioWithVideo = mergeAudioWithVideo;
async function checkForAudioStream(file) {
    ffmpeg.setFfprobePath(settings_1.ffmpegSettings.getFfprobePath());
    return new Promise((resolve, reject) => {
        ffmpeg.ffprobe(file, (err, metadata) => {
            if (err) {
                console.error(`error checking for audioStream for file ${file}`, err);
                reject(err);
                return;
            }
            const audioStreams = metadata.streams.filter(s => s.codec_type === 'audio');
            resolve(audioStreams.length > 0);
        });
    });
}
exports.checkForAudioStream = checkForAudioStream;
async function getSampleRate(filePath) {
    ffmpeg.setFfprobePath(settings_1.ffmpegSettings.getFfprobePath());
    return new Promise((resolve, reject) => {
        ffmpeg.ffprobe(filePath, (err, metadata) => {
            if (err) {
                reject(err);
                return;
            }
            const audioStream = metadata.streams.find(s => s.codec_type === 'audio');
            if (audioStream && audioStream.sample_rate) {
                resolve(audioStream.sample_rate);
            }
            else {
                reject(new Error('No audio stream found'));
            }
        });
    });
}
exports.getSampleRate = getSampleRate;
async function getVideoCodec(filePath) {
    ffmpeg.setFfprobePath(settings_1.ffmpegSettings.getFfprobePath());
    return new Promise((resolve, reject) => {
        ffmpeg.ffprobe(filePath, (err, metadata) => {
            if (err) {
                reject(err);
                return;
            }
            const videoStream = metadata.streams.find(s => s.codec_type === 'video');
            if (videoStream && videoStream.codec_name) {
                resolve(videoStream.codec_name);
            }
            else {
                reject(new Error('No video stream found'));
            }
        });
    });
}
exports.getVideoCodec = getVideoCodec;
