// @revideo/2d/lib/utils/video/ffmpeg-client.js
export class ImageCommunication {
    constructor() {
        this.nextFrameHandlers = [];
        if (!import.meta.hot) {
            throw new Error('FfmpegVideoFrame can only be used with HMR.');
        }
        import.meta.hot.on('revideo:ffmpeg-decoder:video-frame-res', this.handler.bind(this));
    }
    handler(event) {
        const handlers = this.nextFrameHandlers;
        this.nextFrameHandlers = [];
        for (const handler of handlers) {
            handler(event);
        }
    }
    async getFrame(id, src, time, duration, fps) {
        return new Promise((resolve, reject) => {
            if (!import.meta.hot) {
                reject('FfmpegVideoFrame can only be used with HMR.');
                return;
            }
            function handler(event) {
                const image = new Image();
                const uint8Array = new Uint8Array(event.data.frame.data);
                const blob = new Blob([uint8Array], { type: 'image/png' });
                const url = URL.createObjectURL(blob);
                image.src = url;
                image.onload = () => {
                    resolve(image);
                };
            }
            this.nextFrameHandlers.push(handler);
            import.meta.hot.send('revideo:ffmpeg-decoder:video-frame', {
                data: {
                    id: id,
                    filePath: src,
                    startTime: time,
                    duration,
                    fps,
                },
            });
        });
    }
}
