// @revideo/core/lib/exporter/FFmpegExporter.js
var _a;
import { EventDispatcher } from '../events';
import { BoolMetaField, ObjectMetaField } from '../meta';
/**
 * FFmpeg video exporter.
 *
 * @remarks
 * Most of the export logic is handled on the server. This class communicates
 * with the FFmpegBridge through a WebSocket connection which lets it invoke
 * methods on the FFmpegExporterServer class.
 *
 * For example, calling the following method:
 * ```ts
 * async this.invoke('process', 7);
 * ```
 * Will invoke the `process` method on the FFmpegExporterServer class with `7`
 * as the argument. The result of the method will be returned as a Promise.
 *
 * Before any methods can be invoked, the FFmpegExporterServer class must be
 * initialized by invoking `start`.
 */
export class FFmpegExporterClient {
    static meta(project) {
        return new ObjectMetaField(this.displayName, {
            fastStart: new BoolMetaField('fast start', true),
            includeAudio: new BoolMetaField('include audio', true).disable(!project.audio),
        });
    }
    static async create(project, settings) {
        return new _a(project, settings);
    }
    constructor(project, settings) {
        this.project = project;
        this.settings = settings;
    }
    async start() {
        const options = this.settings.exporter.options;
        await this.invoke('start', {
            ...this.settings,
            ...options,
            audio: this.project.audio,
            audioOffset: this.project.meta.shared.audioOffset.get() - this.settings.range[0],
        });
    }
    async handleFrame(canvas) {
        const blob = await new Promise(resolve => canvas.toBlob(resolve, 'image/jpeg'));
        if (!blob) {
            throw Error('Failed to convert canvas to Blob.');
        }
        const dataUrl = await this.blobToDataUrl(blob);
        await this.invoke('handleFrame', {
            data: dataUrl,
        });
    }
    async blobToDataUrl(blob) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();
            reader.onloadend = () => resolve(reader.result);
            reader.onerror = reject;
            reader.readAsDataURL(blob);
        });
    }
    async stop(result) {
        await this.invoke('end', result);
    }
    async kill() {
        await this.invoke('kill', {});
    }
    async generateAudio(assets, startFrame, endFrame) {
        await fetch('/audio-processing/generate-audio', {
            method: 'POST',
            body: JSON.stringify({
                tempDir: `revideo-${this.settings.name}-${this.settings.hiddenFolderId}`,
                assets,
                startFrame,
                endFrame,
                fps: this.settings.fps,
            }),
        });
    }
    async mergeMedia() {
        const outputFilename = this.settings.name;
        const tempDir = `revideo-${this.settings.name}-${this.settings.hiddenFolderId}`;
        await fetch('/audio-processing/merge-media', {
            method: 'POST',
            body: JSON.stringify({
                outputFilename,
                tempDir,
            }),
        });
    }
    /**
     * Remotely invoke a method on the server and wait for a response.
     *
     * @param method - The method name to execute on the server.
     * @param data - The data that will be passed as an argument to the method.
     *               Should be serializable.
     */
    invoke(method, data) {
        if (import.meta.hot) {
            return new Promise((resolve, reject) => {
                const handle = (response) => {
                    if (response.method !== method) {
                        return;
                    }
                    _a.response.unsubscribe(handle);
                    if (response.status === 'success') {
                        resolve(response.data);
                    }
                    else {
                        reject({
                            message: `An error occurred while exporting the video: ${response.message}`,
                            remarks: `Method: ${method}<br>Server error: ${response.message}`,
                            object: data,
                        });
                    }
                };
                _a.response.subscribe(handle);
                import.meta.hot.send('revideo:ffmpeg-exporter', { method, data });
            });
        }
        else {
            throw new Error('FFmpegExporter can only be used locally.');
        }
    }
}
_a = FFmpegExporterClient;
FFmpegExporterClient.id = '@revideo/core/ffmpeg';
FFmpegExporterClient.displayName = 'Video (FFmpeg)';
FFmpegExporterClient.response = new EventDispatcher();
(() => {
    if (import.meta.hot) {
        import.meta.hot.on(`revideo:ffmpeg-exporter-ack`, (response) => _a.response.dispatch(response));
    }
})();
