// @revideo/core/lib/exporter/ImageExporter.js
var _a;
import { FileTypes } from '../app/presets';
import { EventDispatcher } from '../events';
import { BoolMetaField, EnumMetaField, NumberMetaField, ObjectMetaField, } from '../meta';
import { clamp } from '../tweening';
const EXPORT_FRAME_LIMIT = 256;
const EXPORT_RETRY_DELAY = 1000;
/**
 * Image sequence exporter.
 *
 * @internal
 */
export class ImageExporter {
    static meta() {
        const meta = new ObjectMetaField(this.name, {
            fileType: new EnumMetaField('file type', FileTypes),
            quality: new NumberMetaField('quality', 100)
                .setRange(0, 100)
                .describe('A number between 0 and 100 indicating the image quality.'),
            groupByScene: new BoolMetaField('group by scene', false).describe('Group exported images by scene. When checked, separates the sequence into subdirectories for each scene in the project.'),
        });
        meta.fileType.onChanged.subscribe(value => {
            meta.quality.disable(value === 'image/png');
        });
        return meta;
    }
    static async create(project, settings) {
        return new _a(project.logger, settings);
    }
    constructor(logger, settings) {
        this.logger = logger;
        this.settings = settings;
        this.frameLookup = new Set();
        this.handleResponse = ({ frame }) => {
            this.frameLookup.delete(frame);
        };
        const options = settings.exporter.options;
        this.projectName = settings.name;
        this.quality = clamp(0, 1, options.quality / 100);
        this.fileType = options.fileType;
        this.groupByScene = options.groupByScene;
    }
    async start() {
        _a.response.subscribe(this.handleResponse);
    }
    async handleFrame(canvas, frame, sceneFrame, sceneName, signal) {
        if (this.frameLookup.has(frame)) {
            this.logger.warn(`Frame no. ${frame} is already being exported.`);
            return;
        }
        if (import.meta.hot) {
            while (this.frameLookup.size > EXPORT_FRAME_LIMIT) {
                await new Promise(resolve => setTimeout(resolve, EXPORT_RETRY_DELAY));
                if (signal.aborted) {
                    return;
                }
            }
            this.frameLookup.add(frame);
            import.meta.hot.send('revideo:export', {
                frame,
                sceneFrame,
                data: canvas.toDataURL(this.fileType, this.quality),
                mimeType: this.fileType,
                subDirectories: this.groupByScene
                    ? [this.projectName, sceneName]
                    : [this.projectName],
                groupByScene: this.groupByScene,
            });
        }
    }
    async stop() {
        while (this.frameLookup.size > 0) {
            await new Promise(resolve => setTimeout(resolve, EXPORT_RETRY_DELAY));
        }
        _a.response.unsubscribe(this.handleResponse);
    }
}
_a = ImageExporter;
ImageExporter.id = '@revideo/core/image-sequence';
ImageExporter.displayName = 'Image sequence';
ImageExporter.response = new EventDispatcher();
(() => {
    if (import.meta.hot) {
        import.meta.hot.on('revideo:export-ack', response => {
            _a.response.dispatch(response);
        });
    }
})();
